Startup Dialog

ch-sa opened the pull request:
- first prototype with label list and color picker

ch-sa commented:
Hi,

just a super basic prototype without any nice styling so far, just to show this is in progress. ;)

But is still quite dirty at the moment. Maybe it makes sense to have an extra widget for the label inputs.

Feel free to contribute if you have ideas.

![grafik](https://user-images.githubusercontent.com/26072175/199352313-9a69e875-b6e8-4f07-be13-3ce2466e4daa.png)

chingyulin commented on labelCloud/view/gui.py:
```diff
@@ -238,6 +240,13 @@ def __init__(self, control: "Controller") -> None:
+            startup_dialog.save_class_labels()
+        else:
+            sys.exit()
+
```
```suggestion
self.controller.startup(self)
```
The controller needs to start up after the startup dialog is finished to have the updated class labels.

chingyulin replied:
and remove L237

chingyulin reviewed (approved):
Thank you for the PR! The heavy lifting for the startup dialog has been done in this PR. I can continue the work from this. Do feel free to merge this one as you see fit 🙂

chingyulin commented:
Hi @ch-sa,

Few functionalities added:
- remove buttons for each row
- non-empty string validation for label name
- auto increase the label id
- connect the mode in config with the controller
- use QSpinbox for label id to control the inputs

Refactoring - use string enum to represents to modes

<img width="369" alt="image" src="https://user-images.githubusercontent.com/60384727/200168717-9c7e85dd-3894-4c99-8fdf-b87ab57e7988.png">

Questions:
1. I think the method I use to manage buttons to delete rows are quite ugly and I imagine there are some better methods to handle it. Any idea?
2. The layout will change after adding or removing rows. Do you know if there is a way to fix this? I'm not very familiar with the Qt GUI methods.
<img width="369" alt="image" src="https://user-images.githubusercontent.com/60384727/200169053-62e596cd-1bce-44cf-981e-672e907fdbf0.png">
3. Do you know how to do a validation while users click "Save" and pop up an error dialog explaining what are not valid? For examples, checking the label ids, label names, colors are unique.

ch-sa commented:
Looking good already.

I will check the code in the coming days.

I think there should be a way to catch the save signal and only fire it on success.

For the ids I am not sure if it is a good idea to make them editable. Cause when there exist already label definitions and segmented point clouds that use those ids then changing the ids will either lead to mix ups or missing class definitions for already used ids.

I think it might be better to only allow renaming of existing ids.

ch-sa commented on labelCloud/io/labels/config.py:
```diff
@@ -34,7 +34,7 @@ class LabelConfig(object, metaclass=SingletonABCMeta):
         self.classes: List[ClassConfig]
         self.default: int
-        self.type: str
+        self.type: LabelingMode
```
Good point!

ch-sa commented on labelCloud/view/startup_dialog.py:
```diff
@@ -0,0 +1,228 @@
+            label_color = ColorButton(color=rgb_to_hex(class_label.color))
+            # delete button
+            label_delete = QPushButton(text="X")
+            self.delete_buttons.addButton(label_delete, hash(label_delete))
```
I always get the following exception here:
```
self.delete_buttons.addButton(label_delete, hash(label_delete))
OverflowError: argument 2 overflowed: value must be in the range -2147483648 to 2147483647
```

What is the background for using the hashes here?

ch-sa replied:
I think the hash can be at times to high for the id. I changed so that the `button.id` is now directly set as the row index, which also removes the necessity for the hash list.

But otherwise the ButtonGroup seems like a nice fit for this kind of problem.

ch-sa commented:
I had a bit of a fight with PyQt regarding spacing and reading out layouts. While it's still not optimal, I would say it is good enough for now ;)

I changed the following:
1. Added random color for new labels
2. Changed the row identification based on the passed delete button instance
3. Added scrollbar and more spacing (there is now another element in the `class_labels` which made me switch to reasoning on the instance)
4. Added icon and text

**Result**
![welcome-dialog](https://user-images.githubusercontent.com/26072175/201492590-087469db-2fd6-4f17-9121-3b636b797274.gif)

Things that we could still add/ improve (though not necessary for merging):
1. Add icons with bbox / points to the mode buttons for visualization.
2. Add a red border to line edits that fail validation

ch-sa commented:
Merging now, so this PR does not become even bigger.

I would say the last thing missing to merge [feature/semantic-segmentation](https://github.com/ch-sa/labelCloud/tree/feature/semantic-segmentation) on master is the handling of existing labels that are not part of the `_classes.json` yet.

That means whenever a new label file / point cloud is loaded and there exist label classes that are not part of the `_classes.json` yet, they need to be automatically added (with a random color) to that file.

Currently labelCloud simply crashes in this situation. I can do that for object detection. But can you do that for semantic segmentation, @chingyulin?
